pyomo/contrib/cp/repn/docplex_writer.py

```diff
@@ -0,0 +1,1094 @@
+        if arg_scale[-1] is not None:
+            _min, _max, _step = arg_scale[-1]
+            # ESJ: Have to use integer division here because otherwise, later,
+            # when we construct the element constraint, docplex won't believe
+            # the index is an integer expression.
+            if _step is None:
+                raise ValueError(
+                    "Variable indirection '%s' is over a discrete domain "
+                    "without a constant step size. This is not supported."
+                    % node)
+            expr += mult * (arg[1] - _min) // _step
+            # This could be (_max - _min) // _step + 1, but that assumes
+            # that the set correctly collapsed the bounds and that the
+            # lower and upper bounds were part of the step.  That
+            # *should* be the case for Set, but I am suffering from a
+            # crisis of confidence at the moment.
+            mult *= len(arg_domain[-1])
+    # Get the list of all elements selectable by the argument
+    # expression(s); fill in new variables for any indices allowable by
+    # the argument expression(s) but not present in the IndexedComponent
+    # indexing set.
+    elements = []
+    for idx in SetProduct(*arg_domain):
+        try:
+            idx = idx if len(idx) > 1 else idx[0]
+            elements.append(data[0][1][idx])
+        except KeyError:
+            raise ValueError(
+                "Variable indirection '%s' permits an index '%s' "
+                "that is not a valid key. In CP Optimizer, this is a "
+                "structural infeasibility." % (node, idx))
+            # NOTE: If we thought it was the right thing to do in the future, we
+            # could fill in with a bogus variable and add a constraint
+            # disallowing it from being selected
+    try:
+        return (_ELEMENT_CONSTRAINT, cp.element(elements, expr))
+    except:
+        return (_DEFERRED_ELEMENT_CONSTRAINT, (elements, expr))
+
+def _handle_getattr(visitor, node, obj, attr):
+    # We either end up here because we do not yet know the list of variables to
+    # make an element constraint (the first case) or because we are asking for
+    # an attribute on something with indirection, so at this point we *have* a
+    # constructed element constraint (the second case).
+    if obj[0] is _DEFERRED_ELEMENT_CONSTRAINT:
+        # then obj[1] is a list of cp thingies that we need to get the attr on,
+        # and then at the end we need to make the element constraint we couldn't
+        # make before.
+        objects = obj[1][0]
+        ans = []
+        for o in objects:
+            if attr[1] == 'start_time':
+                ans.append(cp.start_of(o))
+            elif attr[1] == 'end_time':
+                ans.append(cp.end_of(o))
+            elif attr[1] == 'length':
+                ans.append(cp.length_of(o))
+            elif attr[1] == 'is_present':
+                ans.append(cp.presence_of(o))
```

Comment:
Can this make use of a `dict` instead of an if tree?  Something like
```python
getattr_dispatcher = {
    'start_time': cp.start_of,
    'end_time': cp.end_of,
    'length': cp.length_of,
    'is_present': cp.presence_of,
}
```
then
```python
ans = list(map(getattr_dispatcher[attr[1]], obj[1][0]))
```

Reply:
Yes, but is there a not-terribly-slow way to catch the `KeyError` and give a useful error message, if I wrap the `ans = ` line in a try-except?

Reply:
I guess I know `attr[1]` but not the specific object, right? Maybe that's enough for now...

Reply:
The dispatcher is independent of the object (and in the code above, the gispatcher is looked up *once* when preparing the arguments to `map()`).  You can always make that more explicit:
```python
try:
    fcn = getattr_dispatcher[attr[1]]
except KeyError:
   # something useful

ans = list(map(fcn, obj[1][0]))
```

There is a slight (25%) performance hit for wrapping something like a dict lookup in a `try-except` ... if you never expect to hit the exception, you could use a defaultdict (but then the message you would get back may not have the context - in particular, not the value of `attr[1]`)